nocasedict/_nocasedict.py

```diff
@@ -37,85 +37,86 @@
 
 
 class NocaseDict(object):
+    # pylint: disable=line-too-long
     """
-    Yet another implementation of a case-insensitive dictionary.
+    A case-insensitive, ordered, and case-preserving, dictionary.
 
-    Whenever keys are looked up, that is done case-insensitively. Whenever
-    keys are returned, they are returned with the lexical case that was
-    originally specified. The dictionary preserves the order of its items
-    (on all supported Python versions).
+    The dictionary is case-insensitive: Whenever items of the dictionary are
+    looked up by key or key values are compared, that is done
+    case-insensitively. The case-insensitivity is defined by performing the
+    lookup or comparison on the result of the ``lower()`` method on the key
+    value. Therefore, objects used as keys must support the ``lower()`` method.
+    If a key object does not do that, :exc:`py:TypeError` is raised.
 
-    Except for the case-insensitivity of its keys, it behaves like the built-in
-    :class:`~py:collections.OrderedDict`.
+    The dictionary is ordered: The dictionary maintains its order based on when
+    the items have been added, for all Python versions supported by this
+    package. This is consistent with the ordering behavior of the built-in
+    :class:`py:dict` class starting with Python 3.7.
 
-    In addition to the methods listed, the dictionary supports:
+    The dictionary is case-preserving: Whenever key values are returned, they
+    have the lexical case that was originally specified when adding or updating
+    the item.
 
-      * Retrieval of values based on key: `val = d[key]`
+    The implementation maintains a dictionary with the lower-cased keys, and
+    items that are a tuple (original key, value).
 
-      * Assigning values for a key: `d[key] = val`
+    Except for the case-insensitivity of its keys, the
+    :class:`~nocasedict.NocaseDict` class behaves like the built-in
+    :class:`py:dict` class starting with Python 3.7 (where it is guaranteed to
+    be ordered), so its documentation applies completely.
```

Comment:
Pending our discussion, doesn't this statement depend on our decision about hash?

Reply:
Yes. We don't need to resolve that now, there is an issue open for that.